### DISCUSSION TITLE
3X3 Man Mesh Compile Error

### DISCUSSION THREAD

**Discussion by Cal-LF** [Jyers/Marlin#839] [General Discussion]
Been trying to compile the 3X3 Man Mesh but get the Fail shown in the photo. Trying the stock folder without copying the config files gets a success when compiled. 

![3X3 Man](https://user-images.githubusercontent.com/76499230/118376578-c634c100-b596-11eb-81d6-37f1552bc560.jpg)

**Comment by Jyers** [comment]
Fixed 😄

**Reply by agalakhov** [reply]
C++ hint: to avoid such errors, don't place big hunks of code under `#ifdef`. Make a function and place only the call to this function under `#ifdef`. The compiler will optimize out all unused functions anyway. If doing so, the compiler will always check even the unused code.

**Reply by Jyers** [reply]
@agalakhov Interesting, that's what I initially was doing, but it seemed the general Marlin format was to put all unused functions in compile time conditionals. Also I believe the compiler will throw a warning if a function is defined but never called. Thoughts?

**Reply by agalakhov** [reply]
The warning can be silenced:
https://en.cppreference.com/w/cpp/language/attributes/maybe_unused
Marlin uses C++17 now, so this is Ok.

Looks like Marlin abuses `#idef`, probably for historical reasons. They just don't consider side effects. Probably the best usage of `#ifdef` (like many other things) is the one of the Linux kernel:
https://www.kernel.org/doc/html/v4.10/process/coding-style.html#conditional-compilation
This style guide needs extra clarification if you didn't read the Linux kernel source code yet. Looks like it gives you multiple choices but it doesn't, there are non-written rules as well. The priorities are:

* If possible, write stub version of the function and use preprocessor to choose:
```
void level_bltouch() {
   // ...
}

void level_none() {
}

#ifdef BLTOUCH
  #define level level_bltouch
#else
  #define level level_none
#endif
```

* If the function in question is not compilable in other configurations, then (and only then) place it under `#ifdef`:
```
#ifdef BLTOUCH
void level_bltouch() {
  // ...
}
#endif
```

* If the operation in question is not doable at the function level, use plain `if` and let the compiler optimize out:
```
if (IS_ENABLED(BLTOUCH) {
  // ...
} else {
  // ...
}
```

* And the `#ifdef` directly in the code is the last resort.

In Marlin we don't have to make it _so_ strict, but possible side effects have to be considered. There are very few people in the world that really know C++, and I can tell you that I don't really know C++ even after 20+ years of coding with it despite the fact I also was a C++ teacher and not a bad one. If there are multiple choices, the choice is not quite arbitrary even if looks so. We have to choose the safest approach.

Another example is the order of `#include` statements. While it could be random, there is good reason to include them in that order: the `*.h` corresponding to the same `*.cpp` first, then other files of the same submodule, then other submodules of the same project, and finally the standard library. But most people do exactly the opposite. Reasoning: if a include statement is forgotten in `*.h`, but the `*.cpp` includes the forgotten header before your one, the code will still compile. Other users of your `*.h` will have mysterious errors.

**Reply by Jyers** [reply]
Incredible... Thanks so much for the great incite! I never would have considered using the preprocessor to define a functions using stubs like that. That's such a great approach! And the include order, I never would have thought of that. Have I mentioned how glad I am you're here? :)
